### PR TITLE
transport/server: add :method POST to incoming metadata

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds a hardcoded :method POST to Server's metadata as per A41 in support of the RBAC HTTP Filter.

The transport verifies and makes sure the only accepted :method is POST.

RELEASE NOTES: None

### CODE REVIEW

File: binarylog/binarylog_end2end_test.go [hunk 1]
```diff
@@ -850,11 +850,25 @@ func equalLogEntry(entries ...*pb.GrpcLogEntry) (equal bool) {
 			tmp := append(h.Metadata.Entry[:0], h.Metadata.Entry...)
 			h.Metadata.Entry = tmp
 			sort.Slice(h.Metadata.Entry, func(i, j int) bool { return h.Metadata.Entry[i].Key < h.Metadata.Entry[j].Key })
+			// Delete headers that have POST values here since we cannot control
+			// this.
+			for i, entry := range h.Metadata.Entry {
+				if entry.Key == ":method" {
+					h.Metadata.Entry = append(h.Metadata.Entry[:i], h.Metadata.Entry[i+1:]...)
```
Comment (menghanl) [inline binarylog/binarylog_end2end_test.go:857]: This is wrong if there are more than one `:method` entry in the list. (It won't happen in practice though).

Either make it break after in the if.
Or `i++` only the key doesn't match `:method`.

File: test/end2end_test.go [hunk 2]
```diff
@@ -7819,3 +7819,36 @@ func (s) TestStreamingServerInterceptorGetsConnection(t *testing.T) {
...
+		return nil, status.Error(codes.NotFound, "metadata was not in context")
+	}
+	method := md.Get(":method")
+	if len(method) != 1 {
+		return nil, status.Error(codes.NotFound, ":method value had more than one value")
+	}
+	if method[0] != "POST" {
+		return nil, status.Error(codes.NotFound, ":method value was not post")
+	}
+	return nil, status.Error(codes.OK, "")
```
Comment (menghanl) [inline test/end2end_test.go:7835]: `return handler(ctx, req)`

File: test/end2end_test.go [hunk 3]
```diff
@@ -7819,3 +7819,36 @@ func (s) TestStreamingServerInterceptorGetsConnection(t *testing.T) {
...
 }
+
+func unaryInterceptorVerifyPost(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, status.Error(codes.NotFound, "metadata was not in context")
+	}
+	method := md.Get(":method")
+	if len(method) != 1 {
+		return nil, status.Error(codes.NotFound, ":method value had more than one value")
```
Comment (menghanl) [inline test/end2end_test.go:7830]: This and the next shouldn't use NotFound.
Use InvalidArgument?

File: test/end2end_test.go [hunk 4]
```diff
@@ -7819,3 +7819,36 @@ func (s) TestStreamingServerInterceptorGetsConnection(t *testing.T) {
...
+	}
+	return nil, status.Error(codes.OK, "")
+}
+
+// TestUnaryInterceptorGetsPost verifies that the server transport adds a
+// :method POST header to metadata, and that that added Header is visibile at
+// the grpc layer.
+func (s) TestUnaryInterceptorGetsPost(t *testing.T) {
+	ss := &stubserver.StubServer{}
+	if err := ss.Start([]grpc.ServerOption{grpc.UnaryInterceptor(unaryInterceptorVerifyPost)}); err != nil {
```
Comment (menghanl) [inline test/end2end_test.go:7843]: You don't need an interceptor. Just read and check the metadata in `StubServer.EmptyCall`.

**Review by menghanl — CHANGES_REQUESTED** [review]

File: test/end2end_test.go (hunk 4 above)
Comment (menghanl) [reply test/end2end_test.go:7843]: Or, actually, it is OK. The purpose is to test interceptors.

Comment (zasweq) [reply test/end2end_test.go:7847]: Yeah, because it's in the context of RBAC. Agree with you otherwise though.

File: binarylog/binarylog_end2end_test.go (hunk 1 above)
Comment (zasweq) [reply binarylog/binarylog_end2end_test.go:857]: Hmmmm I see what you're saying, would delete from list as you're iterating? Broke after deleting :method.

File: test/end2end_test.go (hunk 2 above)
Comment (zasweq) [reply test/end2end_test.go]: Switched.

File: test/end2end_test.go (hunk 3 above)
Comment (zasweq) [reply test/end2end_test.go]: Switched.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :D!

File: binarylog/binarylog_end2end_test.go [hunk 5]
```diff
@@ -850,11 +850,26 @@ func equalLogEntry(entries ...*pb.GrpcLogEntry) (equal bool) {
...
 		}
 		if h := e.GetServerHeader(); h != nil {
 			tmp := append(h.Metadata.Entry[:0], h.Metadata.Entry...)
 			h.Metadata.Entry = tmp
 			sort.Slice(h.Metadata.Entry, func(i, j int) bool { return h.Metadata.Entry[i].Key < h.Metadata.Entry[j].Key })
+			// Delete headers that have POST values here since we cannot control
+			// this.
+			for i, entry := range h.Metadata.Entry {
+				if entry.Key == ":method" {
+					h.Metadata.Entry = append(h.Metadata.Entry[:i], h.Metadata.Entry[i+1:]...)
```
Comment (menghanl) [inline binarylog/binarylog_end2end_test.go:870]: And break here.

Comment (zasweq) [reply binarylog/binarylog_end2end_test.go:870]: Whoops sorry, done.

**Review by zasweq — COMMENTED** [review]
Thanks for the comment :D!

File: internal/transport/http2_server.go [hunk 6]
```diff
@@ -380,6 +380,7 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 			s.recvCompress = hf.Value
 		case ":method":
 			httpMethod = hf.Value
+			mdata[":method"] = append(mdata[":method"], hf.Value)
```
Comment (dfawley) [inline internal/transport/http2_server.go:383]: If we are adding this, I believe we also need to add validation that it _is_ POST.

I thought the plan was to synthesize this to be POST?

Comment (zasweq) [reply internal/transport/http2_server.go:383]: We do have a validation that it is POST. This was actually my first PR merged: https://github.com/grpc/grpc-go/pull/4241.

Comment (zasweq) [reply internal/transport/http2_server.go:383]: See RBAC Filter PR for me and Menghan's back and forth on this topic.

Comment (dfawley) [reply internal/transport/http2_server.go:383]: Whoops.  I had forgotten that.  So you might want to change this in the PR description, then, as it's misleading:

> This PR adds a hardcoded :method POST to Server's metadata

--

> See RBAC Filter PR for me and Menghan's back and forth on this topic.

tl;dr so I don't have to go hunting for the right thread?  This seems pretty straightforward to me.

Comment (menghanl) [reply internal/transport/http2_server.go:383]: I updated the title.

File: binarylog/binarylog_end2end_test.go [hunk 7]
```diff
@@ -850,11 +850,27 @@ func equalLogEntry(entries ...*pb.GrpcLogEntry) (equal bool) {
...
+				}
+				break
```
Comment (menghanl) [inline binarylog/binarylog_end2end_test.go:872]: Hmm, inside the `if`

**Review by menghanl — APPROVED** [review]

File: internal/transport/http2_server.go (hunk 6 above)
Comment (zasweq) [reply internal/transport/http2_server.go:383]: Basically, Eric wants a :method POST in the metadata, because RBAC needs to know this for matching logic. However, the codebase had no previous append to the metadata adding :method POST. The only thing we had was it client side, and reading the value and making sure it was POST, but not in the metadata structure itself. Thus, I decided to put the appending of :method POST in the interceptor, but Menghan mentioned that was hacky and it would be better if I did it in transport, as that is more general solution. Since this broke some tests, I decided to do this in a separate PR.

Comment (dfawley) [reply internal/transport/http2_server.go:383]: Thanks.  SGTM!
